js/test/Exchange/test.trade.js

```diff
@@ -1,93 +1,67 @@
+    assert (trade);
     const sampleTrade = {
-        'info':         { 'a': 1, 'b': 2, 'c': 3 }, // the original decoded JSON as is
-        'id':           '12345-67890:09876/54321',  // string trade id
-        'timestamp':    1502962946216,              // Unix timestamp in milliseconds
-        'datetime':     '2017-08-17 12:42:48.000',  // ISO8601 datetime with milliseconds
-        'symbol':       'ETH/BTC',                  // symbol
-        'order':        '12345-67890:09876/54321',  // string order id or undefined/None/null
-        'type':         'limit',                    // order type, 'market', 'limit' or undefined/None/null
-        'side':         'buy',                      // direction of the trade, 'buy' or 'sell'
-        'takerOrMaker': 'taker',                    // string, 'taker' or 'maker'
-        'price':        0.06917684,                 // float price in quote currency
-        'amount':       1.5,                        // amount of base currency
-        'cost':         0.10376526,                 // total cost (including fees), `price * amount`
-    }
-
-    assert.containsAllKeys (trade, sampleTrade)
-    assert.hasAnyKeys (trade, [ 'fee', 'fees' ])
-
-    const feeKeys = [ 'cost', 'currency' ]
-
-    if (trade.fee !== undefined) {
-        assert (typeof trade.fee === 'object')
-        assert.containsAllKeys (trade.fee, feeKeys)
+        'info': { 'a': 1, 'b': 2, 'c': 3 },    // the original decoded JSON as is
+        'id': '12345-67890:09876/54321',       // string trade id
+        'timestamp': 1502962946216,            // Unix timestamp in milliseconds
+        'datetime': '2017-08-17 12:42:48.000', // ISO8601 datetime with milliseconds
+        'symbol': 'ETH/BTC',                   // symbol
+        'order': '12345-67890:09876/54321',    // string order id or undefined/None/null
+        'type': 'limit',                       // order type, 'market', 'limit' or undefined/None/null
+        'side': 'buy',                         // direction of the trade, 'buy' or 'sell'
+        'takerOrMaker': 'taker',               // string, 'taker' or 'maker'
+        'price': 0.06917684,                   // float price in quote currency
+        'amount': 1.5,                         // amount of base currency
+        'cost': 0.10376526,                    // total cost (including fees), `price * amount`
+    };
+    const keys = Object.keys (sampleTrade);
+    for (let i = 0; i < keys.length; i++) {
+        const key = keys[i];
+        assert (key in trade);
     }
-
-    if (trade.fees !== undefined) {
-        assert (typeof trade.fees === 'object')
-        assert (Array.isArray (trade.fees))
-        trade.fees.forEach ((fee) => assert.containsAllKeys (fee, feeKeys))
+    const fee = ('fee' in trade) ? trade['fee'] : undefined;
+    const fees = ('fees' in trade) ? trade['fees'] : undefined;
+    // logical XOR
+    if (fee || fees) {
+        assert (!(fee && fees));
```

Comment:
We're breaking the logic of the fee test here.
The purpose of it is to make sure:

- either `fee` or `fees` is present
- if it's a `fee` - each field of the `fee` has to be a correct type and value
- if it's `fees` - each of those should be tested for correct types and values in each of the `fees` structs

Reply:
We should probably do this:

- have a test.fee.js and transpile it to `test_fee.php` and `test_fee.py`. 
- have a `test.fees.js` that would reuse the `test.fee.js`
- reuse `test_fee.php` in `test_fees.php
- reuse `test_fee.py` in `test_fees.py`
- reuse `test.fee.js` and `test.fees.js` in `test.trade.js` and `test.order.js`